### ISSUE TITLE
3.5 release next steps: Code freeze Monday 5/17

### ISSUE THREAD

**Issue by hexfusion** [etcd-io/etcd#12970]
etcd 3.5 has seen a tremendous amount of new features bug fixes and performance and stability improvements. At this time I would like to thank everyone for the hard work and dedication. As per the community meeting on May 6th we have outlined Monday, May 17th as code freeze for the 3.5 release. Below are the proposed series of events for input.

Steps:
- [x] cut `release-3.5` branch from master (now main). All future merges to main will be 3.6 moving forward (Monday AM 5/17).
- [x] cut new v3.5.0-beta.3
- [x] cut new modularized release v3.5.0-beta.3 (Monday AM 5/17).
- [x] begin correctness and performance/stress testing (ideas, ownership?)

After testing and validation of the beta, we can consider a sequential beta release or move towards a v3.5 release candidate.

Your input is greatly appreciated.

cc @xiang90 @wenjiaswe @ptabor @gyuho @jingyih @lilic @jpbetz @spzala

**Comment by wenjiaswe** [comment]
Thank you very much @hexfusion ! 

> begin correctness and performance/stress testing (ideas, ownership?)

Would you mind putting together a "gift registry" so we could plan and start pick up?

**Comment by gyuho** [comment]
> cut release-3.5 branch from master (now main). All future merges to main will be 3.6 moving forward (Monday AM 5/17).

As we discussed, @hexfusion @lilic will take ownership on the release management, right?

> begin correctness and performance/stress testing (ideas, ownership?)

We have some in-house component that does this, but I believe our continuous functional tests are good enough.

Please chime in if you have better ideas.

**Comment by gyuho** [comment]
@hexfusion Please feel free to email and ping other maintainers via etcd-dev mailing list.

**Comment by ptabor** [comment]
- [ ] Release bbolt 1.3.6
- [ ] When beta is out we can ask k8s to incorporate https://github.com/kubernetes/kubernetes/pull/100488 (@liggitt ). 
This will allow us to get CI feedback from k8s. 
- [ ] When branch --release-3.5 is tagged, we will run our in-house tests
- [ ] Blog-post was started by @gyuho (https://docs.google.com/document/d/1F2p2LVPo4gj3n8XFMd4QrlarbCdJrbd-pNrT3XLYXmM/edit?usp=sharing)

**Comment by hexfusion** [comment]
> As we discussed, @hexfusion @lilic will take ownership on the release management, right?
correct

**Comment by lilic** [comment]
>  Functional tests are a little flaky (not more than release-3.4), but if that's the main signal, someone should dig in

If no one is looking into this, I can have a look end of this week. @ptabor do you have a list of these? Otherwise I can have a look at previous CI runs.

**Comment by serathius** [comment]
I would be happy to help

**Comment by hexfusion** [comment]
Red Hat team was preparing to cut the release-3.5 branch at approximately 8:00am EDT. If we have a need to hold this please reach out.

cc @ptabor @gyuho @lilic

**Comment by ptabor** [comment]
@hexfusion:  I think we can tag the branch. 
  - I suspect https://github.com/etcd-io/etcd/pull/11427#discussion_r633271675 will require a patch, but its OK. 
  - I still have some hope to merge https://github.com/etcd-io/etcd/pull/12971 into. But this requires your approval.

**Comment by ptabor** [comment]
> > Functional tests are a little flaky (not more than release-3.4), but if that's the main signal, someone should dig in
> 
> If no one is looking into this, I can have a look end of this week. @ptabor do you have a list of these? Otherwise I can have a look at previous CI runs.

I've seen them on CI, but unfortunately I didn't collected the links and its not easy to search by 'jobs' currently. Running in a forloop locally mightbe an option as well.

**Comment by hexfusion** [comment]
> I still have some hope to merge #12971 into. But this requires your approval.
complete

**Comment by hexfusion** [comment]
@ptabor thoughts on remaining nice to have[1]?

[1] https://github.com/etcd-io/etcd/pulls?q=is%3Apr+is%3Aopen+label%3Arel%2F3.5-nice-to-have

**Comment by lilic** [comment]
There is also https://github.com/etcd-io/etcd/issues/12872. But I am not sure we need to do this with code freeze and it can be done afterwards, correct?

**Comment by ptabor** [comment]
I think we are good. 

---
#12933  server: set multiple concurrentReadTx instances share one txReadBuffer  area/performance
   -  seems be very impactful, and I wish it could made eventually into 3.5. 
   - as it touches very critical logic so we are very demanding with testing coverage. 

We might make a call to include this one to beta/beta.2 later. 
---

Small tweaks without public behavior change. We used to incorporate such changes in patch releases. 

#12587 backend: fix buffer range bug  area/bug lgtm 
#12568 backend: set seq flag for each bucket buffer  area/performance 
#12486 server: make raft read only option configurable  

---

Features:  -> probably for 3.6

#10887   etcdserver: add ability to auto-promote learners to voters  rel/3.5-nice-to-have
    - We didn't help the contributor with reviews. Probably too late to start it now. Afraid 3.6.

**Comment by hexfusion** [comment]
@ptabor We are proposing to delay the branch cut until 1:00 EDT to allow the West coast teams final input and a last review of the above. WDYT?

**Comment by ptabor** [comment]
Cutting a brunch alone is not a 'point of no return'. We can decide to backport some PRs from main to `release-3.5` after it happens - just the toil/bar will be a little higher. 

Do you assume cutting a brunch is going atomically with releasing/tagging "etcd-3.5.0-beta.0" or this will be a follow up e.g. tomorrow ?

**Comment by hexfusion** [comment]
My initial thoughts were to create a separation of the beta code by branching. The downside with this approach is that we will need to backport changes. So if we feel there is still a reasonable amount of work to do (appears this way) we could consider holding the branch until we are ready for a release candidate? Thus create a beta release from main this morning.

**Comment by hexfusion** [comment]
The downside of the above is a more manual vetting of merges so it's a trade-off I suppose.

**Comment by ptabor** [comment]
I think we can branch 'release-3.5' today. On our end it will unblock onsite automation that  understands 3.5 to be coming soon. 

If we make a call to backport some changes from 'main' to 'release-3.5' and we consider it too toilish using github PR per change, we can do it in a single PR.

**Comment by serathius** [comment]
> When beta is out we can ask k8s to incorporate kubernetes/kubernetes#100488 (Jordan Liggitt ).
> This will allow us to get CI feedback from k8s.

I can help with that

**Comment by lilic** [comment]
There is also this PR https://github.com/etcd-io/etcd/pull/12979 that seems ready for review and can be merged if we want it.

**Comment by hexfusion** [comment]
@mitake could you possibly take a quick peek at #12979?

**Comment by hexfusion** [comment]
Hi team it is 1:00 EDT as promised I am about to create a new branch for release-3.5 do we have anything blocking this action? I will take any blocking input until quarter past 1:15 PM EDT at which point the branch will be pushed.

cc @xiang90 @wenjiaswe @ptabor @gyuho @jingyih @lilic @jpbetz @spzala

**Comment by hexfusion** [comment]
branch is pushed we will begin in the process of new beta release.

**Comment by hexfusion** [comment]
We have found an upgrade bug[1] that is going to block the beta we expect this to be resolved today. Thanks, @gyuho and @chaochn47 for flushing this out early.

[1] https://github.com/etcd-io/etcd/issues/12987

**Comment by gyuho** [comment]
@hexfusion The fix is ready and tested https://github.com/etcd-io/etcd/pull/12988. We will backport it to 3.5 branch.

**Comment by hexfusion** [comment]
Backport merged we are back on track as `beta.0`[1] was a no-op, the release will be `beta.1`. I will continue the process tonight.

[1] https://github.com/etcd-io/etcd/commit/cf54fab42f072d167a98f2e9ee6168a44f2d191e

**Comment by hexfusion** [comment]
v3.5.0-beta.1 is now available, had some issues with the new modular release going to give a few more tries then will follow up in AM.

[1] https://github.com/etcd-io/etcd/releases/tag/v3.5.0-beta.1

**Comment by liggitt** [comment]
> had some issues with the new modular release going to give a few more tries then will follow up in AM.

once those are available I'll update https://github.com/kubernetes/kubernetes/pull/100488

**Comment by hexfusion** [comment]
@ptabor can you please take a look at v3.5.0-beta.3 just cut I wanted to make sure I understood how you were envisioning each component to look. IE

```
tests/v3.5.0-beta.3
```

**Comment by ptabor** [comment]
@hexfusion Tags looks correct to me. The rules are described here: https://golang.org/ref/mod#vcs-version

```
If a module is defined in a subdirectory within the repository, that is, the module subdirectory portion of the module path is not empty, then each tag name must be prefixed with the module subdirectory, followed by a slash. For example, the module golang.org/x/tools/gopls is defined in the gopls subdirectory of the repository with root path golang.org/x/tools. The version v0.4.0 of that module must have the tag named gopls/v0.4.0 in that repository.
```

**Comment by ptabor** [comment]
The crucial test for me is:
```
ptab@ptab /tmp% cd foobar 

ptab@ptab /tmp/foobar% go mod init test1
go: creating new go.mod: module test1

ptab@ptab /tmp/foobar% go get go.etcd.io/etcd/client/v3
go: downloading go.etcd.io/etcd/client/v3 v3.5.0-beta.3
go: downloading go.etcd.io/etcd/api/v3 v3.5.0-beta.3
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.0-beta.3
go get: added go.etcd.io/etcd/client/v3 v3.5.0-beta.3

ptab@ptab /tmp/foobar% cat go.mod 
module test1

go 1.16

require go.etcd.io/etcd/client/v3 v3.5.0-beta.3 // indirect
```

**Comment by hexfusion** [comment]
OK great so I think we are ok to move forward with `v3.5.0-beta.3`[1]

[1] https://github.com/etcd-io/etcd/releases/tag/v3.5.0-beta.3

cc @liggitt

**Comment by liggitt** [comment]
It looks like this is depending on a pre-release of bbolt (https://github.com/etcd-io/etcd/blob/v3.5.0-beta.3/go.mod#L22). Is that planned to be tagged before this exits beta?

**Comment by liggitt** [comment]
updated https://github.com/kubernetes/kubernetes/pull/100488 with v3.5.0-beta.3

**Comment by hexfusion** [comment]
> It looks like this is depending on a pre-release of bbolt (https://github.com/etcd-io/etcd/blob/v3.5.0-beta.3/go.mod#L22). Is that planned to be tagged before this exits beta?

tracking here https://github.com/etcd-io/bbolt/issues/278

**Comment by chalin** [comment]
Hi @hexfusion et all: I know that y'all are busy with getting the 3.5 release out, but as I mention in #12923 the **text that accompanies the releases needs updating** -- in particular, there shouldn't be links into the website _repo_ (that's an artifact of the days when the docs were in the etcd repo).

If you'd like to see the **reworked text**, see https://github.com/etcd-io/etcd/issues/12923#issuecomment-843641587.

**Comment by mitake** [comment]
@hexfusion sorry for my late response, I reviewed and merged #12979 :)

**Comment by gyuho** [comment]
@hexfusion Any plan for next beta or release candidate? I am also planning to run some benchmarks, but I don't think it's a blocker.

**Comment by hexfusion** [comment]
@gyuho I will cut another beta Tuesday. Then we should talk about cutting bbolt any blockers there on your end? If we can get bbolt and zap we can cut rc0.

I pushed on zap a bit. [1]

cc @ptabor any input?

[1] https://github.com/uber-go/zap/issues/942

**Comment by hexfusion** [comment]
To clarify I will wait until EOD today for backports to clear will cut the release at approx 9PM ET.

**Comment by hexfusion** [comment]
Zap has communicated they will try to cut 1.17 today. This aligns with our goals of an rc release possibly by EOW.

**Comment by ptabor** [comment]
@serathius is driving k8s integration (tests) and internal tests. So far so good *https://github.com/kubernetes/kubernetes/pull/102062)

**Comment by hexfusion** [comment]
[v3.5.0-beta.4](https://github.com/etcd-io/etcd/releases/tag/v3.5.0-beta.4) is now live.

**Comment by nate-double-u** [comment]
I've opened a tracking issue for [v3.5 Documentation](https://github.com/etcd-io/website/issues/328) over in the etcd-io/website repo. Please feel free to add line items/comments to help track the docs changes needed to support the release.

**Comment by ptabor** [comment]
- [ ] https://github.com/etcd-io/etcd/pull/13055 is fixing k8s depending on our integration tests packages (https://github.com/kubernetes/kubernetes/pull/100488#issuecomment-849375418) 
- [ ] https://github.com/etcd-io/etcd/issues/13056 - k8s API chunking test flakes to investigate

**Comment by serathius** [comment]
Found a breaking change in API, https://github.com/etcd-io/etcd/pull/13060

**Comment by serathius** [comment]
Both fixes have been backported to 3.5 branch, we should be ready to cut another beta release.

**Comment by lilic** [comment]
Should we replace the alpha to latest beta in all the go.mod files? Example:

https://github.com/etcd-io/etcd/blob/932d42b0279c21b6e73134729ee7f2831106bc24/go.mod#L23-L32

**Comment by hexfusion** [comment]
> Both fixes have been backported to 3.5 branch, we should be ready to cut another beta release.

Catching up from holidays will cut new release today.

**Comment by gyuho** [comment]
@hexfusion Can we also update bbolt to v1.3.6?

**Comment by hexfusion** [comment]
yes plans are to bump deps and roll rc.0 will get to that tonight.

**Comment by ptabor** [comment]
@lilic @hexfusion 

> Should we replace the alpha to latest beta in all the go.mod files? Example:

>  Can we also update bbolt to v1.3.6?

Both changes and also zap to 1.17.0 waiting in: 
 - https://github.com/etcd-io/etcd/pull/13078
 - https://github.com/etcd-io/etcd/pull/13079 (backport to 3.5)

**Comment by hexfusion** [comment]
both are now merged, thanks for the assist. now pending rc.0

**Comment by hexfusion** [comment]
etcd v3.5.0-rc.0](https://github.com/etcd-io/etcd/releases/tag/v3.5.0-rc.0) is now available for testing.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/100488 updated

**Comment by gyuho** [comment]
@hexfusion Any outstanding item to close out? The target date is June 15, 2021.

**Comment by lilic** [comment]
I believe we need a new release candidate that includes some changes since the 3.5.0-rc.0. Before that we also need this https://github.com/etcd-io/etcd/pull/13049 PR to land in release-3.5 branch.

**Comment by serathius** [comment]
https://github.com/kubernetes/kubernetes/pull/102062 updated

**Comment by hexfusion** [comment]
will cut new rc tonight.

> @hexfusion Any outstanding item to close out? The target date is June 15, 2021.

https://github.com/etcd-io/etcd/pull/13095 is up once merged I can cut rc.1 tonight. Then pending any issues from k8s move towards a formal release.

**Comment by hexfusion** [comment]
etcd [v3.5.0-rc.1](https://github.com/etcd-io/etcd/releases/tag/v3.5.0-rc.1) is now available for testing.

cc @liggitt  @serathius

**Comment by serathius** [comment]
Kubernetes has been soaking v3.5.0-rc.0 server in CI for over a 24h without any issues reported.

**Comment by serathius** [comment]
Kubernetes has been using v3.5.0-rc.0 in their CI for last 24h without any issues reported.

**Comment by hexfusion** [comment]
@gyuho yes nothing blocking at this point shooting for 6PM EDT.

**Comment by gyuho** [comment]
@chalin @nate-double-u Can we make blog post public, right after

> 6PM EDT

?

**Comment by nate-double-u** [comment]
We can do that -- to be clear: blog release June 15, 6PM EDT

Also, I have an open PR for renaming the `next` folder to `v3.5`: https://github.com/etcd-io/website/pull/363

**Comment by nate-double-u** [comment]
(also, please note that @chalin is out this week)

**Comment by mrueg** [comment]
I found these TODOs in the default branch, that might be worth to be cleaned up before release:
```
./server/mvcc/metrics_txn.go:   rangeCounterDebug.Add(ranges) // TODO: remove in 3.5 release
./tests/e2e/v3_curl_test.go:// TODO: remove /v3beta tests in 3.5 release
./tests/e2e/v3_curl_lease_test.go:// TODO remove /kv/lease/timetolive, /kv/lease/revoke, /kv/lease/leases tests in 3.5 release
```

**Comment by nate-double-u** [comment]
Website release prep PR: https://github.com/etcd-io/website/pull/379

**Comment by nate-double-u** [comment]
Standing by. I can merge https://github.com/etcd-io/website/pull/379 when you approve @hexfusion, or when i see v3.5.0 on the [Releases page](https://github.com/etcd-io/etcd/releases).

**Comment by hexfusion** [comment]
Congratulations to everyone for an amazing team effort. etcd team proudly announces GA for  [v3.5.0](https://github.com/etcd-io/etcd/releases/tag/v3.5.0)

**Comment by nate-double-u** [comment]
Here's the blog announcement: https://etcd.io/blog/2021/announcing-etcd-3.5/
Congratulations everyone!
